backend/src/index.ts

```diff
@@ -1,54 +1,94 @@
-import express from 'express';
-
-import { collection, addDoc, getDocs } from 'firebase/firestore';
-import { db } from './firebase';
-import cors from 'cors';
+import express, { Request, Response } from "express";
+import { TypedRequest, LoginBody } from "./requestTypes";
+import bcrypt from "bcrypt";
+import { collection, addDoc, getDocs, where, query } from "firebase/firestore";
+import { db } from "./firebase";
+import cors from "cors";
 
 const EXPRESS_PORT = 3000;
 
 const app = express();
 app.use(express.urlencoded({ extended: true }));
-app.use(express.json())
+app.use(express.json());
 app.use(cors());
 
-app.get('/', (req, res) => {
-  res.send('yellowshirt backend says hello');
+app.get("/", (req, res) => {
+  res.send("yellowshirt backend says hello");
+});
+
+app.post("/register", async (req: TypedRequest<LoginBody>, res: Response) => {
+  const { username, password } = req.body;
+
+  const querySnapshot = await getDocs(collection(db, "users"));
+  if (querySnapshot.docs.some((doc) => doc.data().username === username)) {
+    return res.status(400).send("Username already exists");
+  }
+
+  const saltRounds: number = 10;
+  const hashedPassword: string = await bcrypt.hash(password, saltRounds);
```

Comment:
although bcrypt adds one too, i'd recommend that we add a random [salt](https://en.wikipedia.org/wiki/Salt_(cryptography)) to the password ourselves as well while hashing for extra security. this salt can then be stored in the user obj in the db

Reply:
yes please! this doesnt make a huge difference to our security really - diminishing returns - *but* it's a common practice you'll likely follow when encrypting/hashing with other libraries in the future